### PR TITLE
Improvements on dependencies calculations

### PR DESCRIPTION
**Pull request by Remi-Gau**
#### Metadata treatment
Split `get_metadata` into two functions:

- `metalist = get_meta_list(filename, pattern)` returns list of files satisfying pattern with inheritance principle
- `get_metadata(metafile)` construct meta structure from json files in passed meta-files list (from above function)

Each data file in dataset receives new field `metafile` that contains list of files from `get_meta_list`.

This will facilitate access to metafiles directly. Individual values can be extracted by  user if needed using `get_metadata`.

#### Dependencies

Each file structure contains dependencies sub-structure with guaranteed fields:

- `explicit` -- list of data files containing "IntendedFor" referencing current file.
- `data` -- list of files with same name but different extension. This will combine files that are split in header and data (like in Brainvision), also takes care of bval/bvec files
- `group` -- list of files that have same name except extension and suffix. This will group file that logically need each other, like functional mri and events tabular file. It also takes care of fmap magnitude1/2 and phasediff

Additional custom subfields can be added for given modalities.

This allow to treat `dwi` and `fmap` with generic `parse_using_schema`. `perf` likely also, but I can't certify it.

#### Issues

- [ ] performance. I feel that retrieval of dependencies can be optimised
- [x] Tests -- was unable to install moxunit
- [x] Tests on other datasets

### CODE REVIEW

**Comment by Remi-Gau** [comment]
This should:
- fix #130 
- fix #143 
- partly fix #157 

For qMRI some files in the `derivatives` folder also have an `intendedFor` and they would be missed by the current code.

I suggest we finish this PR as for now and then tackle #146 that concerns the indexing of nested derivative folders before we come back to finish fixing #157 completely.

**Comment by Remi-Gau** [comment]
>     * [ ]  performance. I feel that retrieval of dependencies can be optimised

I suspect that having to load and read the metadata of each file takes a lot of time.

I suspect that this can indeed be optimized but I would worry about that later.

**Comment by Remi-Gau** [comment]
@nbeliy 

I pretty much took your code as is, fixed the test that were not passing and did a bit of refactoring and commenting.

I also:
- did a soft reset to ditch the `derivatives` function from this PR so we can tackle separately in your other PR
- cherry picked a commit that you had on your `copy derivatives` branch to update the query function

The code is pretty much the same though: thanks for providing all this.

Let me know what you think.

Note: I feel that the current way of indexing all the dependencies creates a lot of "redundancies" in the layout (example: the phase image of fieldmap will list all the associated imagesin `dependencies.group` like magnitude and the other way around). 

But I think that we can also optimize this later, if needed as this way of indexing things makes fewer assumptions about what file is / should be the "main" one for a given group.

**Comment by nbeliy** [comment]
> For qMRI some files in the `derivatives` folder also have an `intendedFor` and they would be missed by the current code.

It will still work if `IntendedFor` in `derivatives` reference file in itself, not in the original/main dataset. If it's not, we can imagine a search for files in main dataset at condition that main dataset is referenced in dataset_description.json. But then, it will be difficult to implement, considering that main dataset may be defined as DOI or URL.

> I suspect that having to load and read the metadata of each file takes a lot of time.

Conserving full metadata for each file can explode memory usage. From inside we need only `IntendedFor`, and for user a query with `'target', {'meta1', 'meta2'}` can be imagined to retrieve several parameters with one go. 

> Note: I feel that the current way of indexing all the dependencies creates a lot of "redundancies" in the layout (example: the phase image of fieldmap will list all the associated imagesin dependencies.group like magnitude and the other way around).

Sometimes the main file is not well defined -- in phasediff and 2 magnitude, which one is main? In order to simplify we can exclude `.tsv` files from looking for group dependencies.

File: +bids/+internal/ends_with.m [hunk 1]
```diff
@@ -0,0 +1,23 @@
+function res = ends_with(str, pattern)
+  %
+  % Checks id character array 'str' ends with 'pat'
+  %
+  % USAGE  res = bids.internal.endsWith(str, pat)
+  %
+  % str        - character array
+  % pat        - character array
```
Comment (nbeliy) [inline +bids/+internal/ends_with.m:8]: Will it be a more complexity/performance issue to make it also accept cellstr for pat?

File: +bids/layout.m [hunk 2]
```diff
@@ -182,159 +176,39 @@
...
-      subject = bids.internal.append_to_layout(file_list{i}, subject, modality, schema);
+            subject.(modality)(end).meta = bids.internal.get_metadata(subject.(modality)(iFile).metafile);
 
-      % if this file is a nifti image we add the bval and bvec as dependencies
-      if ~isempty(subject.(modality)) && ...
-              any(strcmp(subject.(modality)(end).ext, {'.nii', '.nii.gz'}))
+            aslcontext_file = strrep(subject.perf(end).filename, ...
+                                     ['_asl' subject.perf(end).ext], ...
+                                     '_aslcontext.tsv');
+            subject.(modality)(end).dependencies.context = manage_tsv( ...
```
Comment (nbeliy) [inline +bids/layout.m:198]: Not sure that asl and aslcontext means, but if it only replaces suffix, will not the `_aslcontext.tsv` appear in group dependencies?

File: +bids/layout.m [hunk 3]
```diff
@@ -452,114 +382,49 @@ function tolerant_message(use_schema, msg)
...
-
-    else
-      intended_for{end + 1, 1} = fullpath_filename;
-
-    end
   end
 
 end
 
 function perf = manage_asllabeling(perf, pth)
```
Comment (nbeliy) [inline +bids/layout.m:425]: This `jpg` image also appears in `group` dependency

File: +bids/query.m [hunk 4]
```diff
@@ -216,11 +219,16 @@
                 result{end + 1} = fullfile(BIDS.subjects(i).path, modalities{j}, d(k).filename);
               end
 
-            case 'metadata'
+            case 'metafiles'
               if isfield(d(k), 'filename')
+                fmeta = BIDS.subjects(i).(modalities{j})(k).metafile;
```
Comment (nbeliy) [inline +bids/query.m:224]: Don't see point of line 223

**Review by nbeliy — APPROVED** [review]
Looks good for mee, more abstract and generic.

As for optimisation, the list of candidates for metadata (considering only json) is same for each call. Probably it can be saved somewhere.

If (to be confirmed) data files are scanned in alphabetical order, lokking for group dependencies can be limited in file's neighbours.

Also, we probably need to move `get_metadata` from internal to util. This function can be usefull outside the layout and query

**Comment by Remi-Gau** [comment]
Will try to adress some of your questions. Before merging.

Totally agree with getting get_metadata out of `internal`, was suggested before and I have been dragging my feet to do it. :man_shrugging: 

Was thinking something along the same line for optimization: will open a separate issue to keep track of that.

File: +bids/query.m (hunk 4 above)
Comment (Remi-Gau) [reply +bids/query.m]: Yup you are correct some of those IF blocks are most likely to always turn out to be true.
Just removed them.

File: +bids/layout.m (hunk 3 above)
Comment (Remi-Gau) [reply +bids/layout.m]: correct

I removed the special way of handling this

File: +bids/+internal/ends_with.m (hunk 1 above)
Comment (Remi-Gau) [reply +bids/+internal/ends_with.m:8]: Should be easy enough: the idea would be to check if the file ends with any pattern from all the patterns in a cell?

Comment (Remi-Gau) [reply +bids/+internal/ends_with.m:8]: FYI: Will most likely move this to a different issue.

File: +bids/layout.m (hunk 2 above)
Comment (Remi-Gau) [reply +bids/layout.m:198]: it will but it is also needed by `layout` to take care of the M0 so at the moment, it might be better to leave it there.

**Pull request merged**